asciidtype/tests/test_asciidtype.py

```diff
@@ -50,24 +51,61 @@ def test_creation_truncation():
 
 
 def test_casting_to_asciidtype():
-    arr = np.array(["hello", "this", "is", "an", "array"], dtype=ASCIIDType(5))
-
-    assert repr(arr.astype(ASCIIDType(7))) == (
-        "array(['hello', 'this', 'is', 'an', 'array'], dtype=ASCIIDType(7))"
-    )
-
-    assert repr(arr.astype(ASCIIDType(5))) == (
-        "array(['hello', 'this', 'is', 'an', 'array'], dtype=ASCIIDType(5))"
-    )
-
-    assert repr(arr.astype(ASCIIDType(4))) == (
-        "array(['hell', 'this', 'is', 'an', 'arra'], dtype=ASCIIDType(4))"
-    )
-
-    assert repr(arr.astype(ASCIIDType(1))) == (
-        "array(['h', 't', 'i', 'a', 'a'], dtype=ASCIIDType(1))"
-    )
-
-    # assert repr(arr.astype(ASCIIDType())) == (
-    #    "array(['', '', '', '', ''], dtype=ASCIIDType(0))"
-    # )
+    for dtype in (None, ASCIIDType(5)):
+        arr = np.array(["this", "is", "an", "array"], dtype=dtype)
+
+        assert repr(arr.astype(ASCIIDType(7))) == (
+            "array(['this', 'is', 'an', 'array'], dtype=ASCIIDType(7))"
+        )
+
+        assert repr(arr.astype(ASCIIDType(5))) == (
+            "array(['this', 'is', 'an', 'array'], dtype=ASCIIDType(5))"
+        )
+
+        assert repr(arr.astype(ASCIIDType(4))) == (
+            "array(['this', 'is', 'an', 'arra'], dtype=ASCIIDType(4))"
+        )
+
+        assert repr(arr.astype(ASCIIDType(1))) == (
+            "array(['t', 'i', 'a', 'a'], dtype=ASCIIDType(1))"
+        )
+
+        # assert repr(arr.astype(ASCIIDType())) == (
+        #    "array(['', '', '', '', ''], dtype=ASCIIDType(0))"
+        # )
+
+
+def test_unicode_to_ascii_to_unicode():
+    arr = np.array(["hello", "this", "is", "an", "array"])
+    ascii_arr = arr.astype(ASCIIDType(5))
+    round_trip_arr = ascii_arr.astype("U5")
+    np.testing.assert_array_equal(arr, round_trip_arr)
+
+
+def test_creation_fails_with_non_ascii_characters():
+    inps = [
+        ["😀", "¡", "©", "ÿ"],
+        ["😀", "hello", "some", "ascii"],
+        ["hello", "some", "ascii", "😀"],
+    ]
+    for inp in inps:
+        with pytest.raises(
+            TypeError,
+            match="Can only store ASCII text in a ASCIIDType array.",
+        ):
+            np.array(inp, dtype=ASCIIDType(5))
+
+
+def test_casting_fails_with_non_ascii_characters():
+    inps = [
+        ["😀", "¡", "©", "ÿ"],
+        ["😀", "hello", "some", "ascii"],
+        ["hello", "some", "ascii", "😀"],
+    ]
+    for inp in inps:
+        arr = np.array(inp)
+        with pytest.raises(
+            TypeError,
+            match="Can only store ASCII text in a ASCIIDType array.",
+        ):
+            arr.astype(ASCIIDType(5))
```

Comment:
I am thinking `astype` already supports `arr.astype(ASCIIDType)`, but I am not sure.  It is definitely reachable e.g. via ufuncs, but that is a bit trickier.

Reply:
It doesn't support it yet, that's https://github.com/numpy/numpy/issues/22756. I'm going to try poking at that once I finish here.

Reply:
Ah, right... I didn't change it anywhere, I guess.  Starting with `astype()` is probably good since it is likely clearer or at least simpler than `np.array()`.